### DISCUSSION TITLE
Guidance: logdensity function with extra arguments

### DISCUSSION THREAD

**Discussion by johannahaffner** [blackjax-devs/blackjax#708] [Q&A]
Hi! 

I'm new to Blackjax, and really appreciate the flexibility it promises. 
I'm currently thinking about how to implement my first use case. 

**Can the logdensity function have a signature other than taking a state and spitting out a value?** 
I understand that this is not possible during the `step`, at least not in the current implementation of `Update_Fn`. In the NUTS sampler, it also does not seem to be possible to pass such arguments when building the kernel.
Is it possible for any algorithm in Blackjax?

I have a generalised-least-squares expression to compute the loglikelihood for the parameters of an ODE model, given the data and a parametric variance model. I already use it as an objective function during parameter estimation, and I now want to sample around the estimate. Ideally, I would like to be able to `vmap` over the data, which represent individual time-series.

From what I have understood so far, the most straightforward thing for me to do would be writing a callable wrapper class for the loglikelihood expression I have and make the other arguments attributes of the wrapper, such that I get an interface that only takes the parameters.

However, this would mean that I need to build the kernel repeatedly (hundreds of times). Will this have a strong impact on the performance of my program?

**Comment by johannahaffner** [comment]
I found https://github.com/blackjax-devs/blackjax/discussions/416, is the use of inner functions the recommended way to create such a wrapper?

In that example, they are being used to create a log density that only depends on the sampled parameter.

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

Yes — the recommended pattern is a closure (or `functools.partial`) that captures the extra arguments and exposes the `position -> logdensity` signature the kernel expects:

```python
logdensity_fn = lambda position: full_logdensity(position, data=data, hyper=hyper)
```

If the extra arg changes per step (e.g. a fresh PRNG key for a stochastic density), build the partial inside your step loop. See #416 and #545 for worked examples.
